build/xtask/src/main.rs

```diff
@@ -159,6 +164,71 @@ struct Config {
     config: Option<ordered_toml::Value>,
 }
 
+#[derive(Clone, Debug)]
+struct Config {
+    name: String,
+    target: String,
+    board: String,
+    signing: IndexMap<String, Signing>,
+    secure: Option<bool>,
+    stacksize: Option<u32>,
+    bootloader: Option<Bootloader>,
+    kernel: Kernel,
+    outputs: IndexMap<String, Output>,
+    tasks: IndexMap<String, Task>,
+    peripherals: IndexMap<String, Peripheral>,
+    extratext: IndexMap<String, Peripheral>,
+    supervisor: Option<Supervisor>,
+    config: Option<ordered_toml::Value>,
+    buildhash: u64,
+}
+
+impl Config {
+    pub fn from_file(cfg: &Path) -> Result<Self> {
+        let cfg_contents = std::fs::read(&cfg)?;
+        let toml: RawConfig = toml::from_slice(&cfg_contents)?;
+
+        let mut hasher = DefaultHasher::new();
+        hasher.write(&cfg_contents);
+
+        // If the app.toml specifies a `chip` key, then load the peripheral
+        // register map from a separate file and accumulate that file in the
+        // buildhash.
+        let peripherals = if let Some(chip) = toml.chip {
+            if !toml.peripherals.is_empty() {
+                bail!("Cannot specify both chip and peripherals");
+            }
+            let mut src_dir = cfg.to_path_buf();
```

Comment:
FYI - `cfg.parent().join(chip)`

Reply:
... `cfg.parent().unwrap().join(chip)` I think. The `src_dir.pop()` implies the equivalent of that `unwrap` if I'm reading correctly.

Reply:
Cool!